### ISSUE TITLE
Provide a full implementation of LLVM IR

### ISSUE THREAD

**Issue by mewmew** [llir/llvm#7]
Based on the time constraints of this project a full implementation of LLVM IR will not be developed during its time frame. The ambition is to develop a full implementation once the project is finished. With this in mind, the focus is now to implement a minimal subset required for decompilation. Any code not directly related to this subset will be removed from the repository for now, and will be added back once the project is completed. This issue will make sure to track these code changes so they can be reverted easily.

**Comment by mewmew** [comment]
Based on time constraints the focus of the project has shifted to finishing up the artifacts (LLVM IR -> DOT [ll2dot](https://github.com/mewrev/ll2dot) and the `iso` and `merge` [tools](https://github.com/mewrev/graphs) which locates and merges subgraph isomorphism in DOT files) which for now uses the official Go binding for LLVM. Future work will continue to develop the pure Go library for interacting with LLVM IR.

[The following comment](https://github.com/mewpaper/decompilation/issues/6#issuecomment-76158157) describes this further.
> Marked as future ambition. Developing a library for interacting with LLVM IR turned out to require huge efforts. This work is taking place in the [llvm](https://github.com/mewlang/llvm) repository and will continue to do so. As the LLVM IR library was a prerequisite for the other parts of this project a decision was made to make use of the preexisting [official Go binding for LLVM](https://godoc.org/llvm.org/llvm/bindings/go/llvm). Using this library the [ll2dot](https://github.com/mewrev/ll2dot) tool was constructed to translate LLVM IR into control flow graphs described in the DOT format. Subsequent decompilation passes worked on the DOT representation of the Control Flow Graph and were therefore independent of the source intermediate representation (e.g. LLVM IR).
> 
> The intention is to take up work again and implement the pure Go implementation of LLVM IR as time permits. The immediate focus of the project is currently to finish the development of the artifact and shift focus to writing the report. As such the development of the LLVM IR library may remain stale until later this year.

**Comment by mewmew** [comment]
This work is tracked by the [llir/spec](https://github.com/llir/spec) project, which aims to define a formal grammar for LLVM IR. As the grammar supports an ever larger subset of LLVM IR, the corresponding in-memory representations are added to this repository. Closing this issue for now, as the [spec](https://github.com/llir/spec) repository is better suited to track this work.

**Comment by mewmew** [comment]
The `llir/spec` project has now been integrated with the `llir/llvm` project. As such, this issue is once again relevant for tracking this work.

Mental note, commit d299b42 removed visibility, linkage and calling conventions from the LLVM IR. These may be included once more in a future version of the llir/llvm package.

**Comment by mewmew** [comment]
ref: https://github.com/llir/llvm/issues/16#issuecomment-264116665

[fast-math-flags](http://llvm.org/docs/LangRef.html#fast-math-flags)
- [ ] nnan
- [ ] ninf
- [ ] nsz
- [ ] arcp
- [ ] fast

**Comment by mewmew** [comment]
This issue is mostly a duplicate of #15. Lets close this one to keep the discussion in one place.
